src/inbox/mod_inbox.erl

```diff
@@ -85,17 +98,23 @@ stop(Host) ->
                  To :: jid:jid(),
                  Acc :: mongoose_acc:t(),
                  IQ :: jlib:iq()) -> {stop, mongoose_acc:t()} | {mongoose_acc:t(), jlib:iq()}.
-process_iq(_From, _To, Acc, #iq{type = set, sub_el = SubEl} = IQ) ->
-    {Acc, IQ#iq{type = error, sub_el = [SubEl, mongoose_xmpp_errors:not_allowed()]}};
-process_iq(From, To, Acc, #iq{type = get, sub_el = QueryEl} = IQ) ->
+process_iq(_From, _To, Acc, #iq{type = get, sub_el = SubEl} = IQ) ->
+    Form = build_inbox_form(),
+    SubElWithForm = SubEl#xmlel{ children = [Form] },
+    {Acc, IQ#iq{type = result, sub_el = SubElWithForm}};
+process_iq(From, To, Acc, #iq{type = set, id = QueryId, sub_el = QueryEl} = IQ) ->
     Username = From#jid.luser,
     Host = From#jid.lserver,
-    List = mod_inbox_backend:get_inbox(Username, Host),
-    QueryId = exml_query:attr(QueryEl, <<"queryid">>, <<>>),
-    forward_messages(List, QueryId, To),
-    BinCount = integer_to_binary(length(List)),
-    Res = IQ#iq{type = result, sub_el = [build_result_iq(BinCount)]},
-    {Acc, Res}.
+    case form_to_params(exml_query:subelement_with_ns(QueryEl, ?NS_XDATA)) of
+        {error, bad_request} ->
+            {Acc, IQ#iq{ type = error, sub_el = [ mongoose_xmpp_errors:bad_request() ] }};
```

Comment:
I know that from security reasons returning just `bad_request` is ok. On the other hand, from my perspective, when client devs tries to adopt custom extensions (even if there is a doc how do to it). They got frustrated not knowing why there is a `bad_request` error. I suggest to extend the error with a text description saying, at least, in which field there was an issue.

Reply:
👍